File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/HomeIntelligenceValidation.kt
@file:Suppress("MagicNumber")

package app.pantopus.android.ui.screens.homes

import app.pantopus.android.data.api.models.homedashboard.HomeHealthScoreDto
import app.pantopus.android.data.api.models.homedashboard.HomePropertyValueDto
import app.pantopus.android.data.api.models.homedashboard.SeasonalChecklistDto
import app.pantopus.android.data.api.models.homedashboard.SeasonalChecklistItemDto
import java.time.OffsetDateTime
import kotlin.math.roundToInt

/** A decoded object is not yet a trustworthy card or mutation confirmation. */
internal object HomeIntelligenceValidation {
    private val dimensions =
        mapOf("maintenance" to 25, "bills" to 20, "seasonal" to 20, "emergency" to 15, "household" to 10, "documents" to 10)
    private val destinations = setOf("maintenance", "bills", "dashboard", "emergency", "members", "documents")
    private val statuses = setOf("pending", "completed", "skipped", "hired")
    private val sources = setOf("cache", "cache_stale", "attom", "fallback", "unavailable")

    fun health(
        value: HomeHealthScoreDto,
        homeId: String,
    ): Boolean {
        if (!healthDimensions(value) || value.topIssue?.isBlank() == true) return false
        val action = value.topAction ?: return true
        return action.type == "navigate" && action.label.isNotBlank() && destinations.any { action.route == "/homes/$homeId/$it" }
    }

    private fun healthDimensions(value: HomeHealthScoreDto): Boolean {
        if (value.breakdown.keys != dimensions.keys || value.score !in 0..100) return false
        if (!value.breakdown.all {
                    (key, part) ->
                part.max == dimensions[key] && part.score in 0..part.max && part.issues.all { it.isNotBlank() }
            }
        ) {
            return false
        }
        return value.score == value.breakdown.values.sumOf { it.score }
    }

    fun checklist(
        value: SeasonalChecklistDto,
        homeId: String,
    ): Boolean {
        if (!checklistRows(value, homeId)) return false
        val all = value.items + value.carryover?.items.orEmpty()
        if (all.map { it.id }.distinct().size != all.size) return false
        val completed = value.items.count { it.isResolved }
        val percentage = if (value.items.isEmpty()) 0 else (100.0 * completed / value.items.size).roundToInt()
        return value.progress.total == value.items.size && value.progress.completed == completed && value.progress.percentage == percentage
    }

    private fun checklistRows(
        value: SeasonalChecklistDto,
        homeId: String,
    ): Boolean {
        if (value.season.key.isBlank() || value.season.label.isBlank()) return false
        if (!value.items.all { item(it, homeId) && it.seasonKey == value.season.key }) return false
        val carryover = value.carryover
        if (carryover != null) {
            return carryover.season.key.isNotBlank() && carryover.season.label.isNotBlank() &&
                carryover.items.all { item(it, homeId) && it.seasonKey == carryover.season.key && it.status == "pending" }
        }
        return true
    }

    fun item(
        value: SeasonalChecklistItemDto,
        homeId: String,
    ): Boolean =
        value.homeId == homeId && value.id.isNotBlank() && value.title.isNotBlank() &&
            !value.seasonKey.isNullOrBlank() && value.year?.let { it in 1..9999 } == true && !value.itemKey.isNullOrBlank() &&
            value.status in statuses && value.sortOrder >= 0 && timestamp(value.completedAt) &&
            (value.status != "hired" || !value.gigId.isNullOrBlank())

    fun property(value: HomePropertyValueDto): Boolean {
        if (!listOf(
                value.estimatedValue,
                value.valueRangeLow,
                value.valueRangeHigh,
            ).all { it == null || it.isFinite() && it > 0 }
        ) {
            return false
        }
        return propertyMetadata(value) && propertyRange(value)
    }

    private fun propertyMetadata(value: HomePropertyValueDto): Boolean {
        if (value.source !in sources || !timestamp(value.lastUpdated)) return false
        if (value.valueConfidence?.let { !it.isFinite() || it !in 0.0..100.0 } == true) return false
        if (value.zipMedianSalePriceTrend?.let { it !in setOf("up", "down", "flat") } == true) return false
        return value.yearBuilt?.let { it in 1..9999 } != false && value.sqft?.let { it > 0 } != false
    }

    private fun propertyRange(value: HomePropertyValueDto): Boolean {
        val estimate = value.estimatedValue
        val low = value.valueRangeLow
        val high = value.valueRangeHigh
        if (value.source == "unavailable" && listOf(estimate, low, high, value.valueConfidence).any { it != null }) return false
        if (low != null && high != null && low > high) return false
        if (estimate != null && low != null && estimate < low) return false
        return estimate == null || high == null || estimate <= high
    }

    private fun timestamp(value: String?): Boolean = value == null || runCatching { OffsetDateTime.parse(value) }.isSuccess
}

File: frontend/apps/ios/Pantopus/Features/Homes/HomeIntelligenceValidation.swift
import Foundation

/// Decoding alone does not establish a complete card or an exact Home receipt.
enum HomeIntelligenceValidation {
    private static let dimensions = ["maintenance": 25, "bills": 20, "seasonal": 20, "emergency": 15, "household": 10, "documents": 10]
    private static let destinations = ["maintenance", "bills", "dashboard", "emergency", "members", "documents"]
    private static let statuses = ["pending", "completed", "skipped", "hired"]
    private static let sources = ["cache", "cache_stale", "attom", "fallback", "unavailable"]

    static func health(_ value: HomeHealthScoreDTO, homeId: String) -> Bool {
        guard Set(value.breakdown.keys) == Set(dimensions.keys), (0...100).contains(value.score) else { return false }
        guard value.breakdown.allSatisfy({ key, part in
            part.max == dimensions[key] && (0...part.max).contains(part.score) && part.issues.allSatisfy(nonblank)
        }) else { return false }
        guard value.score == value.breakdown.values.reduce(0, { $0 + $1.score }) else { return false }
        if let issue = value.topIssue, !nonblank(issue) { return false }
        guard let action = value.topAction else { return true }
        return action.type == "navigate" && nonblank(action.label) && destinations.contains { action.route == "/homes/\(homeId)/\($0)" }
    }

    static func checklist(_ value: SeasonalChecklistDTO, homeId: String) -> Bool {
        guard nonblank(value.season.key), nonblank(value.season.label) else { return false }
        guard value.items.allSatisfy({ item($0, homeId: homeId) && $0.seasonKey == value.season.key }) else { return false }
        if let carryover = value.carryover {
            guard nonblank(carryover.season.key), nonblank(carryover.season.label) else { return false }
            guard carryover.items
                .allSatisfy({ item($0, homeId: homeId) && $0.seasonKey == carryover.season.key && $0.status == "pending" })
            else { return false }
        }
        let all = value.items + (value.carryover?.items ?? [])
        guard Set(all.map(\.id)).count == all.count else { return false }
        let completed = value.items.filter(\.isResolved).count
        let percentage = value.items.isEmpty ? 0 : Int((100.0 * Double(completed) / Double(value.items.count)).rounded())
        return value.progress.total == value.items.count && value.progress.completed == completed && value.progress.percentage == percentage
    }

    static func item(_ value: SeasonalChecklistItemDTO, homeId: String) -> Bool {
        guard value.homeId == homeId, nonblank(value.id), nonblank(value.title),
              let season = value.seasonKey, nonblank(season), let year = value.year, (1...9999).contains(year),
              let key = value.itemKey, nonblank(key), statuses.contains(value.status), value.sortOrder >= 0,
              timestamp(value.completedAt) else { return false }
        return value.status != "hired" || value.gigId.map(nonblank) == true
    }

    static func property(_ value: HomePropertyValueDTO) -> Bool {
        guard let source = value.source, sources.contains(source), timestamp(value.lastUpdated) else { return false }
        guard [value.estimatedValue, value.valueRangeLow, value.valueRangeHigh].compactMap({ $0 }).allSatisfy({ $0.isFinite && $0 > 0 })
        else { return false }
        if let confidence = value.valueConfidence, !confidence.isFinite || !(0...100).contains(confidence) { return false }
        if let trend = value.zipMedianSalePriceTrend, !["up", "down", "flat"].contains(trend) { return false }
        if let year = value.yearBuilt, !(1...9999).contains(year) { return false }
        if let sqft = value.sqft, sqft <= 0 { return false }
        return propertyRange(value)
    }

    private static func propertyRange(_ value: HomePropertyValueDTO) -> Bool {
        if value.source == "unavailable",
           [value.estimatedValue, value.valueRangeLow, value.valueRangeHigh, value.valueConfidence]
           .contains(where: { $0 != nil }) { return false }
        if let low = value.valueRangeLow, let high = value.valueRangeHigh, low > high { return false }
        if let estimate = value.estimatedValue, let low = value.valueRangeLow, estimate < low { return false }
        if let estimate = value.estimatedValue, let high = value.valueRangeHigh, estimate > high { return false }
        return true
    }

    private static func timestamp(_ value: String?) -> Bool {
        guard let value else { return true }
        let formatter = ISO8601DateFormatter()
        if formatter.date(from: value) != nil { return true }
        formatter.formatOptions.insert(.withFractionalSeconds)
        return formatter.date(from: value) != nil
    }

    private static func nonblank(_ value: String) -> Bool {
        !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}
